backends/tfhe-cuda-backend/cuda/src/crypto/torus.cuh

```diff
@@ -129,6 +150,121 @@ __host__ void host_modulus_switch_inplace(cudaStream_t stream,
+                                  Torus *output, const Torus *input,
+                                  uint32_t size, uint32_t log_modulus) {
+  cuda_set_device(gpu_index);
+
+  int num_threads = 0, num_blocks = 0;
+  getNumBlocksAndThreads(size, 1024, num_blocks, num_threads);
+
+  modulus_switch<Torus><<<num_blocks, num_threads, 0, stream>>>(
+      output, input, size, log_modulus);
+  check_cuda_error(cudaGetLastError());
+}
+
+template <typename T>
+__device__ __forceinline__ T round_error(T input, uint32_t log_modulus) {
+  T rounded;
+  constexpr uint32_t BITS = sizeof(T) * 8;
+  modulus_switch<T>(input, rounded, log_modulus);
+  rounded <<= (BITS - log_modulus);
+  rounded -= input;
+  return rounded;
+}
+
+// This method is based on rust's
+// core_crypto::centered_binary_ms_body_correction_to_add()
+template <typename T>
+__device__ T centered_binary_modulus_switch_body_correction_to_add(
+    const T *lwe, const uint32_t lwe_dimension, const uint32_t log_modulus) {
+  T sum_half_mask_round_errors = 0;
+  signed_torus_t<T> sum_halving_errors_doubled = 0;
+  constexpr auto TWO = static_cast<signed_torus_t<T>>(2);
+
+  for (auto i = 0; i < lwe_dimension; i++) {
+    auto error = round_error(lwe[i], log_modulus);
+    auto signed_error = static_cast<signed_torus_t<T>>(error);
+    auto half_error = signed_error / TWO;
+
+    // Dividing by 2 can add an error where |error| <= 1/2 in each run of the
+    // loop Combined, they can add up to more than 1 (in the mod 2^64 torus)
+    // Thus we compute this combined error to reduce it to less than 1/2
+    // half_error = half_error_theoretical + halving_error_doubled/2
+    // where half_error_theoretical * 2 = signed_error
+    auto halving_error_doubled = (half_error * TWO) - signed_error;
+
+    sum_half_mask_round_errors += static_cast<T>(half_error);
+    sum_halving_errors_doubled += halving_error_doubled;
+  }
+
+  auto sum_halving_errors = static_cast<T>(sum_halving_errors_doubled / TWO);
+
+  // sum(half_error_theoretical) = sum(half_error) -
+  // sum(halving_error_doubled)/2
+  sum_half_mask_round_errors -= sum_halving_errors;
+
+  constexpr uint32_t BITS = sizeof(T) * 8;
+  auto half_case = static_cast<T>(1) << (BITS - log_modulus - 1);
+
+  // E(e_MMS) = - sum(mask_round_error / 2)
+  // body_centered = body_input - E(e_MMS) - half_case
+  // body_centered = body_input + sum(mask_round_error / 2) - half_case
+  // body_correction_to_add = sum(mask_round_error / 2) - half_case
+  return sum_half_mask_round_errors - half_case;
+}
+
+template <typename Torus>
+__global__ void centered_modulus_switch(Torus *output, const Torus *input,
+                                        uint32_t lwe_dimension,
+                                        uint32_t log_modulus) {
+  const int tid = threadIdx.x + blockIdx.x * blockDim.x;
+
+  if (tid == lwe_dimension) {
+    Torus ee_mms = centered_binary_modulus_switch_body_correction_to_add(
```

Comment:
The result here is actually `-E(e_mms)`, so the name is misleading
Adding it to body as done here is correct

Could be named `correction` instead

Comment applies to other places

Reply:
Indeed!
I will replace it by `correction` everywhere.